### DISCUSSION TITLE
[On hold] Make the library lovely to use

### DISCUSSION THREAD

**Discussion by Mikolaj** [Mikolaj/mostly-harmless#12] [Ideas]
Edit: Let's wait until we decide how to add vector/array operations to the engine, but that technical decision is going to influence the UI profoundly.

--

Simon says:

> MK's current approach is a combination of writing things in monadic style (less convenient for the programmer) and using type classes for dual numbers (very convenient, but with tricky restrictions around boxed values).  An alternative would be to write a Core-to-Core pass as a GHC plugin.

Apart of a GHC plugin, we could also consider a shallowly or deeply embedded DSL, with various ways of coping with the Monad, e.g., with a judicious use of `unsafePerformIO` or C FFI hacks (perhaps similar to what library ad does) to hide it completely from the user.

I think this is very much orthogonal to the question of how to extend our formalism to be able to compete with library backprop on MNIST. But it's somehow related to the issue of generalizing the type of input parameters.

**Comment by simonpj** [comment]
An advantage of the Core-to-Core approach is that all the problems with unboxed things would go away, I think.  And if we want special cases for `Array Float`, say,  we can just do it, rather than having to rely on (say) overlappping instance declarations or whatnot.   More control.

I'm not sure whether the `unsafePerformIO` and C FFI hacks can be made completely benign and invisible ---- or can they?

**Reply by Mikolaj** [reply]
[Edit: let me take advantage of the threading in discussion and move my standalone comments here:]

If the Core-to-Core pass could, e.g., turn every boxed vector of `Double` and of `Float` and it's operations to unboxed ones, that could indeed help at least with the `Traversable` instance. Ideally, the pass would also represent vectors of pairs (that is, dual numbers) as pairs  of vectors (an unboxed vector of scalars and a boxed vector of deltas) and tweak operations on them accordingly. That should be doable, but it's hard to do in Haskell due to types, especially for monadic operations, where it seems it can't be done efficiently via `ST`, but requires `IO`. I wonder, however, if the operations, before or after the tweaks, would fuse nicely with other `vector` package operations, if it turns out it matters (e.g., `hmatrix` is based on `Data.Vector.Unboxed.Vector`, so it may matter).

Anyway, right now I don't have much to say, because I'm confused about what extensions we really need for the library not to be a toy, or at best a teaching tool, and how completely the extensions would up-end the current API.

**Reply by Mikolaj** [reply]
For context, as seen in the benchmarks I published, training with our library is 30 times slower on common sizes of MNIST neural networks (5e5 = 0.5 * 10^6 parameters; papers detailing experiments with such sizes were first published in 1998) than using a hadwritten gradient with hmatrix operations and 20 times slower than when also deriving roughly the same gradient using library backprop. I'd guess our memory usage compares similarly. Looking at how the runtime scales, the proportion may be worse for larger networks because, in our case, deriving the gradient dominates the cost and scales with the size of the matrices.

Edit: however, I'm totally ignorant of potential uses other than ML. If we could come up with any, it'd be great to benchmark.

Edit: we are now faster than handwritten gradient, but the design is still in flux, so IMHO too early for polish.

**Comment by Mikolaj** [comment]
A this point, as a user of the library that already implemented basic recurrent neural networks and now thinks about convolutional networks to implement, I can tell that "lovely" for me would be type-level sizes of vectors and matrices. It's starting to take forever to debug mismatched sizes of matrices and vectors to multiply, etc. Tracking the source of runtime errors in GHC is fine, but lapack sometimes doesn't report what the mismatched sizes are (e.g., when zipping vectors). Also, with types many constraints can be considered at once and so errors are reported early, often closer to the original point where the human made a mistake.

Edit: Lennart's https://hackage.haskell.org/package/orthotope and https://hackage.haskell.org/package/orthotope-hmatrix look like a great model (ideally, after some extensions, a ready solution) for the optionally dependently typed tensors.
